File: Week_03/permutations_ii/Solution.java
package permutations_ii;

import java.util.*;

/**
 * 回溯，在叶子节点判断重复
 * Time: O(n * n!)
 */
class Solution1 {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, res, new ArrayList<>(), new boolean[nums.length], new HashSet<>());
        return res;
    }
    int count = 0;
    private void dfs(int[] nums, List<List<Integer>> res, List<Integer> list, boolean[] used, Set<String> resultCheck) {
        if (list.size() == nums.length) {
            String listStr = list.toString();
            if (!resultCheck.contains(listStr)) {
                res.add(new ArrayList<>(list));
                resultCheck.add(listStr);
            }
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            System.out.println(count++);
            list.add(nums[i]);
            used[i] = true;
            dfs(nums, res, list, used,resultCheck);
            list.remove(list.size() - 1);
            used[i] = false;
        }
    }
    public static void main(String[] args) {
        List<List<Integer>> list = new Solution1().permuteUnique(new int[]{1, 1, 2});
        System.out.println(list);
    }
}

/**
 * 回溯 + 剪枝
 */
class Solution2 {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);//剪枝条件的前提
        dfs(nums, res, new ArrayList<>(), new boolean[nums.length]);
        return res;
    }

    private void dfs(int[] nums, List<List<Integer>> res, List<Integer> list, boolean[] used) {
        if (list.size() == nums.length) {
            res.add(new ArrayList<>(list));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;
            list.add(nums[i]);
            used[i] = true;
            dfs(nums, res, list, used);
            list.remove(list.size() - 1);
            used[i] = false;
        }
    }

    public static void main(String[] args) {
        List<List<Integer>> list = new Solution2().permuteUnique(new int[]{1, 1, 2, 2});
        System.out.println(list);
    }
}

File: Week_03/permutations_ii/Solution.py
# 回溯 + 剪枝
# 剪枝条件：i > 0 and nums[i] == nums[i - 1] and not used[i - 1] 去掉重复元素放在同一个格子的分支
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def dfs_permute(path = []):
            if len(path) == n:
                res.append(path[:])
                return
            for i in range(n):
                if used[i] or i > 0 and nums[i] == nums[i - 1] and not used[i - 1]: continue

                path.append(nums[i])
                used[i] = True
                dfs_permute(path)

                path.pop()
                used[i] = False

        n = len(nums)
        res = []
        used = [False] * n
        nums.sort() #剪枝的前提
        dfs_permute()
        return res